Chain elementStart/elementEnd instructions

crisbeto opened the pull request:
Split up into two commits to make it easier to review:

1. **refactor(compiler): rework instruction generation logic for improved flexibility**
Previously the logic for generating chained instructions was somewhat rigid, because we had to collect all of the calls ahead of time and then call one of the chained instruction helpers. This doesn't work for something like `elementStart`, because we have to descend into other elements that could add to the chain.

These changes refactor the code so that we collect the list of instructions in a flat array and we do the chaining only once at the end when we have the entire instruction set for the code block.

The new approach has the advantage of being (almost) entirely configuration-based via the `CHAINABLE_INSTRUCTIONS` array and being more flexible in allowing us to chain instructions that span across elements.

2. **perf(compiler): chain element start/end instructions** 

In templates with several levels of nested nodes, it's common for several `elementStart`/`elementEnd` instructions to show up in a row which can be optimized away.

These changes add chaining support for `elementStart`, `elementEnd`, `elementContainerStart` and `elementContainerEnd` to shave off some bytes when possible.

crisbeto committed:
refactor(compiler): rework instruction generation logic for improved flexibility

Previously the logic for generating chained instructions was somewhat rigid, because we had to collect all of the calls ahead of time and then call one of the chained instruction helpers. This doesn't work for something like `elementStart`, because we have to descend into other elements that could add to the chain.

These changes refactor the code so that we collect the list of instructions in a flat array and we do the chaining only once at the end when we have the entire instruction set for the code block.

The new approach has the advantage of being (almost) entirely configuration-based via the `CHAINABLE_INSTRUCTIONS` array and being more flexible in allowing us to chain instructions that span across elements.

alfaproject commented:
OOC don't compression algorithms already take care of this?

Can it be opt-in at least? I think for debugging the ability to step on each call is easier with the current way

crisbeto commented:
A while ago when we first started chaining instructions, we had a testing setup that showed it compressing slightly better with the chained instructions. Unfortunately, I haven't been able to dig it up, but a quick test against the `interpolation_nested_context.js` example showed that the template is 2 bytes smaller after gzipping. As for debugging, hasn't this been the case already? The majority of instructions have been chained for a long time now.

crisbeto commented:
Caretaker note: I'll run a TGP for this tomorrow. It moves around a bunch of code in the template builder and I want to double-check that there are no regressions.

AndrewKushnir commented on packages/compiler/src/render3/view/util.ts:
```diff
@@ -50,6 +50,81 @@ export const NON_BINDABLE_ATTR = 'ngNonBindable';
 
+/** Instructions that support chaining. */
+const CHAINABLE_INSTRUCTIONS = new Set([
+  R3.elementStart,
```
Can we chain `element` and `elementContainer` (self-closing) as well?

crisbeto replied:
Done. I didn't do it initially, because it seemed like something that wouldn't happen so often.

AndrewKushnir reviewed (approved):
👍

crisbeto committed:
perf(compiler): chain element start/end instructions

In templates with several levels of nested nodes, it's common for several `elementStart`/`elementEnd` instructions to show up in a row which can be optimized away.

These changes add chaining support for `elementStart`, `elementEnd`, `elementContainerStart` and `elementContainerEnd` to shave off some bytes when possible.

crisbeto commented:
Passes TGP, marking as ready to merge.

dylhunn commented:
This PR was merged into the repository by commit 1b91e1049eedfb96febd9f6cc45fb7c0e1f154a7.

mihaipanait commented:
this caused a lot of pain

TypeError: t.TgZ(...) is not a function
TgZ being ɵɵelementStart()

![2022-02-24 15_12_31-localhost_4200 and 4 more pages - Personal - Microsoft​ Edge](https://user-images.githubusercontent.com/3032687/155589262-81a77e94-9fd9-40c6-8dfd-c9b8b2f397d7.png)

the compiler produced something like this, but the ɵɵelementStart was returning null:

![2022-02-24 16_06_21-localhost_4200 and 20 more pages - Personal - Microsoft​ Edge](https://user-images.githubusercontent.com/3032687/155589821-a96d1288-3fee-49b1-bcab-8e29ccfcb3a7.png)

it was fixed after this change:

![2022-02-24 20_56_02-Sourcetree](https://user-images.githubusercontent.com/3032687/155589943-5492a6bc-1c6c-4f69-a15e-d4918e2301d0.png)

this seems like a breaking change

JoostK commented:
> this seems like a breaking change

@mihaipanait Angular framework packages must be exactly in sync with each other, any other configuration is not supported.
